posts/2020-03-10-calculating-pi-concurrently.adoc

```diff
@@ -0,0 +1,926 @@
+
+[source,scala]
+----
+  val piF: Future[BigDecimal] = indexes
+    .logAtInfo("pi-index")
+    .mapAsync(Settings.parallelism)(piBBPdeaTermIF)
+    .runWith(sumOfTerms)
+----
+
+== Using Substreams to introduce concurrency
+
+An alternative way to speed-up our calculation is to utilise https://doc.akka.io/docs/akka/current/stream/stream-substream.html[Akka Substreams].
+One can consider Substreams as a way to de-multiplex a stream of elements. Substreams
+can be created in different ways, but we'll focus on the `groupBy` combinator.
+`groupBy` takes two arguments, let's start with the second one: this is a function `f`
+that takes an element and which returns a key. The key will determine to which
+Substream the element will be sent. The idea is that `f` returns a finite number
+of unique key values. The first parameter `maxSubstreams` is the number of distinct
+key values generated by the function. Note that if `f` generates more distinct key
+values than `maxSubstreams`, the stream will fail. The following code segment shows
+the splitting of our main stream into Substreams:
+
+[source,scala]
+----
+  val piF = indexes
+    .groupBy(Settings.parallelism, _ % Settings.parallelism)
+----
+
+Next, we can perform the calculation of the terms in each Substream by mapping
+over each index. Also, we can calculate the sum of all the terms in each Substream:
+
+[source,scala]
+----
+  val piF = indexes
+      .groupBy(Settings.parallelism, _ % Settings.parallelism)
+      .map(piBBPdeaTermI).async
+      .fold(BigDecimal(0)){case (acc, term) => acc + term}
+----
+
+With this, we will have `Settings.parallelism` Substreams that each generate one
+`BigDecimal` value. In order to calculate the total sum of the terms, we should
+merge the Substreams into one Stream. We can do this with the `mergeSubstreams`
+combinator. We can complete the calculation in the same manner as in the previous
+solution.
+
+The stream processing now looks as follows:
+
+[source,scala]
+----
+  val piF: Future[BigDecimal] = indexes
+    .groupBy(Settings.parallelism, _ % Settings.parallelism)
+    .map(piBBPdeaTermI)
+    .fold(BigDecimal(0)){case (acc, term) => acc + term}
+    .mergeSubstreams
+    .runWith(sumOfTerms)
+----
+
+An important note about the `mergeSubstreams` combinator is that it takes
+elements from the Substreams as they arrive. This means in our case is that the
+order in which the subtotals are added isn't deterministic.
+
+When we run this code, we observe that, in terms of performance, we're back to
+square one. This is because our Blueprint is optimised, fused and run on a
+single actor. We can fix this by adding an asynchronous boundary in the right
+location like shown in the final version of the stream processing:
+
+[source,scala]
+----
+  val piF: Future[BigDecimal] = indexes
+    .groupBy(Settings.parallelism, _ % Settings.parallelism)
+    .map(piBBPdeaTermI).async
+    .fold(BigDecimal(0)){case (acc, term) => acc + term}
```

Comment:
In this version the per-group summing is done in parallel to `piBBPdeaTermI`, because it's in a substream, isn't it? That's different from the implementation based on `Future`s?

Reply:
You're correct, and that is probably worth mentioning